### ISSUE TITLE
CUDA on different branches

### ISSUE THREAD

**Issue by danielmccusker** [jvzonlab/OrganoidTracker#3]
I see that on the master branch, installing cuda and cuDNN are taken care of now by anaconda. If I want to work on the stable branch and have installed the dependencies from the dependency file on that branch, in the current version of anaconda, do I need to use the old installation instructions to install cuda and cuDNN as well?

**Comment by danielmccusker** [comment]
To be more precise, the graphics card I am running is newer and doesn't support cuda 9.0. The newer versions of cuda only seem to be compatible with tensorflow-gpu >2.0, and I am worried that the code in the stable branch will not run with a newer version of tensorflow. Do you have any suggestions for managing these dependencies in order to use the stable branch?

**Comment by rutgerkok** [comment]
The code on the stable branch will indeed not run on Tensorflow 2, which is kinda annoying. If you want to get the code on the stable branch working, there are probably multiple ways to do it. What I would suggest is to first create a Conda environment with all the other packages for OrganoidTracker, and then try to install Tensorflow 1.15 in it (the most recent version of Tensorflow 1).

Installing the other packages works like this:

    conda create -n organoid_tracker -c chaubold -c conda-forge  python=3.6 dpct matplotlib-base mahotas tifffile scikit-image scipy nd2reader opencv pyside2

Here on Stackoverflow people seem to found a way to install Tensorflow 1.15 within the Conda environment: https://stackoverflow.com/a/64820487

**Comment by danielmccusker** [comment]
Nice, the CUDA installer gave me an issue about CUDA 9.0 and my driver, but using conda to directly install cudatoolkit 9 and cudnn 7 into the environment specified by the win spec file on the stable branch seems to have worked.

Could you advise on best practices for training the neural network? I am generating training data by manually specifying positions of nuclei, but in one of the manuals, it indicates that the software should draw a box, within which every unlabelled pixel is presumably given the label "not nucleus center". In the edit mode I can add nuclear positions but I don't know how to make this box appear. 

I am working with time series of images and different experiments have different delta t, e.g. one experiment has something like 3 min and another something like 2 min. Is this a problem for the flow solver and making predictions for new data, or is it taken care of if I correctly specify dt for each time lapse?

One more thing, how do I add labels for cell divisions? If I understand correctly, I should label 1. Cell positions 2. Cell tracks 3. Cell divisions and then train each network separately.

**Comment by danielmccusker** [comment]
With respect to the first question, we have some manually corrected TrackMate output tracks that I want to use as labeled data. How does the network handle all of the tracks in a time lapse that are unlabeled, i.e. tracks that TrackMate didn't find or the ones that we manually removed?

**Comment by rutgerkok** [comment]
> Could you advise on best practices for training the neural network? I am generating training data by manually specifying positions of nuclei, but in one of the manuals, it indicates that the software should draw a box, within which every unlabelled pixel is presumably given the label "not nucleus center". In the edit mode I can add nuclear positions but I don't know how to make this box appear.

The box isn't actually drawn anywhere, you'd need to imagine this box yourself. In the neural network, it will look at hthe min and max x and y of the labels, and then train only within that area. ( see http://github.com/jvzonlab/OrganoidTracker/blob/c97816db97acbe306e82a3c11162f4c12453d184/organoid_tracker/position_detection_cnn/training_dataset.py#L133 )

> I am working with time series of images and different experiments have different delta t, e.g. one experiment has something like 3 min and another something like 2 min. Is this a problem for the flow solver and making predictions for new data, or is it taken care of if I correctly specify dt for each time lapse?

The flow solver just tries to minimize position and volume changes of nuclei, so it shouldn't matter too much. You can have a look at the organoid_tracker.ini config file that the flow solver will create to adjust the scoring system a bit, if the default parameters don't work well for you.

> With respect to the first question, we have some manually corrected TrackMate output tracks that I want to use as labeled data. How does the network handle all of the tracks in a time lapse that are unlabeled, i.e. tracks that TrackMate didn't find or the ones that we manually removed?

If there is unlabeled data outside the area where there are labeled nuclei, it's not a problem as the network doesn't train there. Howver, if there is unlabeled data inside the area, then this will confuse the network.

In the example image, only the cells with the orange star are labeled. The unlabeled cell with the letter P will be problematic, the other unlabeled cells won't be.

![afbeelding](https://user-images.githubusercontent.com/1462188/185390119-72dec53a-4a2d-457d-a328-5420c7178593.png)

I can also recommend having a look at https://doi.org/10.7554/elife.69380 , it works better with sparsely labeled data and integrates nicely with ImageJ. The 2020 version of OrganoidTracker (the one on the stable branch) uses some hardcoded assumptions on what your nuclei should look like for measuring their volume. I'm not sure if I can still recommend using that version nowadays, now that there are other cell trackers out there with similar approaches, but fewer assumptions on your data.

We are working on a newer version where all steps (nucleus detection, division detection and link detection) are based on deep learning, while still using the flow solver for linking over time. But for now that verson is still work-in-progress.

**Comment by danielmccusker** [comment]
Thanks for the tips. What else does the 2020 version assume about the data? That the nuclei have a particular average size?

**Comment by rutgerkok** [comment]
The size isn't really a problem, the tracker isn't so sensitive to that. If your cells get too small, then detecting them becomes more difficult, and if they get too big you'll need to train using bigger crops. You can change these values in the config. If your nuclei vary a lot in size (like in C Elegans embryos) then you cannot choose good values, and you're out of luck.

Almost all assumptions come from the Gaussian fit. There, the nuclei need to be bright and on a dark background with not too much noise. The nuclei need to also have relatively uniform brightness. Multiple nuclei need to have some darker pixels in between them, so that the Gaussian fit only needs to fit a few cells at once.
